mp2-v1/tests/common/cases/table_source.rs

```diff
@@ -1426,151 +700,265 @@ impl SingleStructExtractionArgs {
+                            current_struct
+                                .update_contract_single_values(ctx, contract)
+                                .await;
+                        }
+                    }
+                    UpdateType::SecondaryIndex => {
+                        if let Some(index_slot_input) = index_slot_input {
+                            let slot = index_slot_input.slot();
+                            let rng = &mut StdRng::from_entropy();
+                            if slot == SINGLE_STRUCT_SLOT as u8 {
+                                let mut current_struct =
+                                    LargeStruct::current_contract_single_values(ctx, contract)
+                                        .await;
+                                let field_index =
+                                    LargeStruct::slot_inputs(SINGLE_STRUCT_SLOT as u8)
+                                        .iter()
+                                        .position(|slot_input| slot_input == &index_slot_input)
+                                        .unwrap();
+                                if field_index == 0 {
+                                    current_struct.field1 += U256::from(1);
+                                } else if field_index == 1 {
+                                    current_struct.field2 += 1;
+                                } else if field_index == 2 {
+                                    current_struct.field3 += 1;
+                                } else {
+                                    panic!("Wrong Struct field index");
+                                }
+                                current_struct
+                                    .update_contract_single_values(ctx, contract)
+                                    .await;
+                            } else {
+                                let mut current_values =
+                                    SimpleSingleValues::current_contract_single_values(
+                                        ctx, contract,
+                                    )
+                                    .await;
+                                if slot == SINGLE_SLOTS[0] {
+                                    current_values.s1 = !current_values.s1;
+                                } else if slot == SINGLE_SLOTS[1] {
+                                    current_values.s2 += U256::from(1);
+                                } else if slot == SINGLE_SLOTS[2] {
+                                    current_values.s3 = Alphanumeric.sample_string(rng, 10);
+                                } else if slot == SINGLE_SLOTS[3] {
+                                    current_values.s4 = next_address();
+                                } else {
+                                    panic!("Wrong slot number");
+                                }
+                                current_values
+                                    .update_contract_single_values(ctx, contract)
```

Comment:
since the updates are _deterministic_ then couldn't we put that inside the trait implementation directly ? Instead of computing the updates outside of the trait (which requires us to be very specific about which struct / value etc it is) and then just ask the trait to update.  Wdyt ? Maybe i'm missing stg obvious here, not sure..